beacon-chain/forkchoice/doubly-linked-tree/forkchoice.go

```diff
@@ -64,19 +65,105 @@ func (f *ForkChoice) Head(
 		return [32]byte{}, errors.Wrap(err, "could not update best descendant")
 	}
+	if err := f.UpdateSafeHead(ctx); err != nil {
+		log.WithError(err).Error("could not update safe head")
+	}
+
 	return f.store.head(ctx)
 }
 
+// UpdateSafeHead updates the safe head in the fork choice store.
+func (f *ForkChoice) UpdateSafeHead(
+	ctx context.Context,
+) error {
+	oldSafeHeadRoot := f.store.safeHeadRoot
+	newSafeHeadRoot, err := f.store.safeHead(ctx)
+	if err != nil {
+		return errors.Wrap(err, "could not get safe head")
+	}
+
+	// If the safe head has not changed, return early.
+	if oldSafeHeadRoot == newSafeHeadRoot {
+		return nil
+	}
+
+	// Update safe head
+	f.store.safeHeadRoot = newSafeHeadRoot
+
+	f.logSafeHead(ctx, newSafeHeadRoot, oldSafeHeadRoot)
+
+	return nil
+}
+
+func (f *ForkChoice) logSafeHead(ctx context.Context, newSafeHeadRoot [32]byte, oldSafeHeadRoot [32]byte) {
+	newSafeHeadNode, ok := f.store.nodeByRoot[newSafeHeadRoot]
+	if !ok || newSafeHeadNode == nil {
+		log.WithError(ErrNilNode).Error("could not find new safe head node")
+		return
+	}
+	newSafeHeadSlot := newSafeHeadNode.slot
+	currentSlot := slots.CurrentSlot(f.store.genesisTime)
+	secondsSinceSlotStart, err := slots.SecondsSinceSlotStart(currentSlot, f.store.genesisTime, uint64(time.Now().Unix()))
+	if err != nil {
+		log.WithError(err).Error("could not compute seconds since slot start")
+	}
+	log.WithFields(logrus.Fields{
+		"currentSlot":        fmt.Sprintf("%d", currentSlot),
+		"sinceSlotStartTime": fmt.Sprintf("%d", secondsSinceSlotStart),
+		"newSafeHeadSlot":    fmt.Sprintf("%d", newSafeHeadSlot),
+		"newSafeHeadRoot":    fmt.Sprintf("%#x", newSafeHeadRoot),
+		"weight":             fmt.Sprintf("%d", newSafeHeadNode.weight),
+	}).Info("Safe head has changed")
+
+	// Update metrics.
+	safeHeadSlotNumber.Set(float64(newSafeHeadSlot))
+
+	// Check if the safe head reorged.
+	commonRoot, forkSlot, err := f.CommonAncestor(ctx, oldSafeHeadRoot, newSafeHeadRoot)
+	if err != nil {
+		log.WithError(err).Error("Could not find common ancestor root")
+		return
+	}
+
+	if oldSafeHeadRoot != [32]byte{} && commonRoot != oldSafeHeadRoot {
+		// The safe head has reorged.
+		// Calculate reorg metrics.
+		oldSafeHeadNode, ok := f.store.nodeByRoot[oldSafeHeadRoot]
+		if !ok || oldSafeHeadNode == nil {
+			log.WithError(ErrNilNode).Error("could not find old safe head node")
+			return
+		}
+		oldSafeHeadSlot := oldSafeHeadNode.slot
+		dis := oldSafeHeadSlot + newSafeHeadSlot - 2*forkSlot
+		dep := max(uint64(oldSafeHeadSlot-forkSlot), uint64(newSafeHeadSlot-forkSlot))
+		log.WithFields(logrus.Fields{
+			"oldSafeHeadSlot":    fmt.Sprintf("%d", oldSafeHeadSlot),
+			"oldSafeHeadRoot":    fmt.Sprintf("%#x", oldSafeHeadRoot),
+			"commonAncestorRoot": fmt.Sprintf("%#x", commonRoot),
+			"distance":           dis,
+			"depth":              dep,
+		}).Debug("Safe head reorg occurred")
```

Comment:
This should not be a debug message but rather a big red warning, possibly a fatal error